### PR TITLE
Update how too-long endpoint hostnames are handled

### PR DESCRIPTION
**Pull request by amisevsk**
### What does this PR do?
Update the hostname format used for endpoints whose hostnames are too long from
```
  <workspace-id>-<order>.<base-domain>
```
to
```
  <workspace-id>-<endpoint-name>.<base-domain>
```
This is necessary as the iteration order through endpoints is random (iterating through Go maps is random), resulting in inconsistent numbers used for `<order>`.

Using a combination of workspace ID and endpoint name should always be valid:

* Workspace IDs are 25 characters long ([ref](https://github.com/devfile/devworkspace-operator/blob/21aaf77d61f8801fbfad940b607f2e9e420b49dd/controllers/workspace/devworkspace_controller.go#L646-L668))
* Endpoint names are restricted to max 15 characters by the Devfile API ([ref](https://devfile.io/docs/2.2.0/devfile-schema#components-container-endpoints-name))
* Endpoint names and workspace IDs are required to be alphanumeric with dashes, starting and ending with an alphanumeric character ([ref](https://devfile.io/docs/2.2.0/devfile-schema#components-container-endpoints-name))
* Endpoint names are unique across all endpoints in the workspace ([ref](https://devfile.io/docs/2.2.1/devfile-validation-rules#endpoints))

### Screenshot/screencast of this PR
N/A

### What issues does this PR fix or reference?
Closes https://github.com/eclipse/che/issues/22774

### How to test this PR?
See reproducer in linked issue; after change, long endpoints should get the hostname
```
workspaceID-endpointname.baseDomain
```

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Development resource are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-development-resources)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: controllers/devworkspace/solver/endpoint_strategy.go [hunk 1]
```diff
@@ -72,8 +72,10 @@ func (u UsernameWkspName) getEndpointPathPrefix(endpointPath string) string {
 func (u UsernameWkspName) getHostname(endpointInfo *EndpointInfo, baseDomain string) string {
 	subDomain := fmt.Sprintf("%s-%s-%s", u.username, u.workspaceName, endpointInfo.endpointName)
 	if errs := validation.IsValidLabelValue(subDomain); len(errs) > 0 {
-		// if subdomain is not valid, use legacy paths
-		return fmt.Sprintf("%s-%d.%s", u.workspaceID, endpointInfo.order, baseDomain)
+		// If subdomain is not valid (e.g. too long), use alternate format
+		// The below should always be under 63 characters, as endpoint names are limited to 15 characters and workspace IDs are
+		// 25 characters.
+		return fmt.Sprintf("%s-%s.%s", u.workspaceID, endpointInfo.endpointName, baseDomain)
```
Comment (tolusha) [inline controllers/devworkspace/solver/endpoint_strategy.go:78]: What about sorting endpoints by theirs names here [1] instead of changing host names?
I just have some concerns, maybe someone relies on hostnames. 

[1] https://github.com/eclipse-che/che-operator/blob/7c7bd4f796eb083a21cf5ba5f750cde1d69e901a/controllers/devworkspace/solver/che_routing.go#L485-L486

Comment (amisevsk) [reply controllers/devworkspace/solver/endpoint_strategy.go:78]: I considered sorting the endpoints instead of this approach, but honestly iterating through a map in sorted order in Go is annoying (since there's no way to easily get a keyset, etc.) :) 

These hostnames are not going to be reliably static in any case since this functionality is triggered only when the default endpoint name becomes too long. For example, some users might see the normal format while others see the legacy format (depending on how long usernames are). You could even end up in a situation where half your endpoints are normal, and half are legacy. In the past, we've pointed people at the `endpoint-url` attribute to determine the actual hostname used for each devfile endpoint.

I'm fine with sorting the endpoints, though I had a small concern that I was missing something in how they're handled, as the [original issue](https://github.com/eclipse/che/issues/22667) has `order` numbers going up to 13 (which may just be how many endpoints the devfile has, I don't know). 

An additional bonus of doing it this way is that we don't run into situations where endpoints could be updated unexpectedly -- e.g. with order, even sorted, if you update the devfile to have an endpoint earlier in the order, all your legacy-hostname endpoints might change.

Comment (tolusha) [reply controllers/devworkspace/solver/endpoint_strategy.go:78]: Thank you for explanation, make sense to me.

**Review by tolusha — APPROVED** [review]

**Review by dkwon17 — APPROVED** [review]

**Comment by amisevsk** [comment]
I took some extra time to test edge cases here:

* Endpoints that expose the same port (allowed if they're on the same container) still end up with unique names
* Devfiles with non-unique endpoint names cannot be started in Che as they fail validation

**Comment by tolusha** [comment]
@amisevsk 
Please ignore `ci/prow/v14-che-behind-proxy` check.
